### PR TITLE
Implement internal/info/info.go benchmark

### PR DESCRIPTION
**Pull request by kevindiu**
### Description:

This PR implements internal/info benchmark.
The result is as follow:
```
goos: darwin
goarch: amd64
pkg: github.com/vdaas/vald/internal/info
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkString-12                	  288968	      5127 ns/op	    7152 B/op	     132 allocs/op
BenchmarkGet-12                   	 1000000	      1056 ns/op	    1714 B/op	      23 allocs/op
Benchmark_info_String-12          	  256093	      6106 ns/op	    5874 B/op	     117 allocs/op
Benchmark_Detail_String-12        	  419288	      2851 ns/op	    4144 B/op	      94 allocs/op
Benchmark_info_Get-12             	 1630857	       764.7 ns/op	    1329 B/op	      16 allocs/op
Benchmark_info_prepare-12         	1000000000	         0.4188 ns/op	       0 B/op	       0 allocs/op
Benchmark_StackTrace_String-12    	 9006242	       138.6 ns/op	     248 B/op	       5 allocs/op
PASS
```

### Environment:

- Go Version: 1.16
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [x] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/info/info_bench_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
+package info
+
+import (
+	"testing"
+)
+
+var stringResult string
+var getResult Detail
```
Comment (vankichi) [inline internal/info/info_bench_test.go:8]:
```suggestion
```

File: internal/info/info_bench_test.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+func initBench() {
+	Init("benchmark_test")
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:12]: Why did you add this?

File: internal/info/info_bench_test.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	var r string
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			r = String()
+		}
+	})
+	stringResult = r
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:26]:
```suggestion
func BenchmarkString(b *testing.B) {
	initBench()
	var r string

	b.ReportAllocs()
	b.ResetTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			String()
		}
	})
}
```

File: internal/info/info_bench_test.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	var r Detail
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			r = Get()
+		}
+	})
+	getResult = r
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:40]: plz fix like as above comment

File: internal/info/info_bench_test.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	var r string
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			r = i.String()
+		}
+	})
+	stringResult = r
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:57]: plz fix like as above comment

File: internal/info/info_bench_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	var r string
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			r = d.String()
+		}
+	})
+	stringResult = r
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:75]: plz fix like as above comment

File: internal/info/info_bench_test.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	var r Detail
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			r = i.Get()
+		}
+	})
+	getResult = r
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:92]: plz fix like as above comment

File: internal/info/info_bench_test.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	in := i.(*info)
+
+	b.ReportAllocs()
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			in.prepare()
+		}
+	})
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:108]: plz fix like as above comment

File: internal/info/info_bench_test.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	var s string
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			s = st.String()
+		}
+	})
+	stringResult = s
+}
```
Comment (vankichi) [inline internal/info/info_bench_test.go:126]: plz fix like as above comment

File: internal/info/info_bench_test.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
+package info
+
+import (
+	"testing"
+)
+
+var getResult Detail
```
Comment (kevindiu) [inline internal/info/info_bench_test.go:7]:
```suggestion
```

File: internal/info/info_bench_test.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
+package info
+
+import (
+	"testing"
+)
+
+func BenchmarkString(b *testing.B) {
+	Init("benchmark")
```
Comment (hlts2) [inline internal/info/info_bench_test.go:8]: Should we check nil or not nil of `infoProvider ` just in case?

File: internal/info/info_bench_test.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if s := String(); s == "" {
+				b.Error("String return empty string")
+			}
+		}
+	})
```
Comment (hlts2) [inline internal/info/info_bench_test.go:17]: Should we clear sync.once method with teardown? 🤔 
I think `BenchmarkString` also use the once method.

File: internal/info/info_bench_test.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+func BenchmarkGet(b *testing.B) {
+	sn := "benchmark"
+	Init(sn)
```
Comment (hlts2) [inline internal/info/info_bench_test.go:22]: same
https://github.com/vdaas/vald/pull/1093/files#r600168339

File: internal/info/info_bench_test.go (hunk 11 above)
Comment (kevindiu) [reply internal/info/info_bench_test.go:8]: I think it is not needed as it should be covered in unit test.

File: internal/info/info_bench_test.go (hunk 13 above)
Comment (kevindiu) [reply internal/info/info_bench_test.go:22]: replied on that thread

File: internal/info/info_bench_test.go (hunk 12 above)
Comment (kevindiu) [reply internal/info/info_bench_test.go:17]: I think it is not needed. This is the benchmark for string method, there is nothing to do with sync.once

File: internal/info/info_bench_test.go (hunk 11 above)
Comment (hlts2) [reply internal/info/info_bench_test.go:8]: I got it. thank you 🙏

**Review by hlts2 — DISMISSED** [review]
LGTM

File: internal/info/info_bench_test.go [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if s := i.String(); s == "" {
+				b.Error("String return empty string")
+			}
+		}
+	})
+}
+
+func Benchmark_Detail_String(b *testing.B) {
```
Comment (vankichi) [inline internal/info/info_bench_test.go:51]: This naming is not correct.

File: internal/info/info_bench_test.go [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			in.prepare()
+		}
+	})
+}
+
+func Benchmark_StackTrace_String(b *testing.B) {
```
Comment (vankichi) [inline internal/info/info_bench_test.go:103]: This naming is not correct.

File: internal/info/info_bench_test.go (hunk 14 above)
Comment (kevindiu) [reply internal/info/info_bench_test.go:51]:
```suggestion
func BenchmarkDetail_String(b *testing.B) {
```

File: internal/info/info_bench_test.go (hunk 15 above)
Comment (kevindiu) [reply internal/info/info_bench_test.go:103]:
```suggestion
func BenchmarkStackTrace_String(b *testing.B) {
```

File: internal/info/info_bench_test.go [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+	b.ReportAllocs()
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			in.prepare()
+		}
+	})
+}
+
+func BenchmarkStackTrace_String(b *testing.B) {
```
Comment (vankichi) [inline internal/info/info_bench_test.go:103]:
```suggestion
func Benchmark_stackTrace_String(b *testing.B) {
```

File: internal/info/info_bench_test.go [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if s := i.String(); s == "" {
+				b.Error("String return empty string")
+			}
+		}
+	})
+}
+
+func BenchmarkDetail_String(b *testing.B) {
```
Comment (vankichi) [inline internal/info/info_bench_test.go:51]:
```suggestion
func Benchmark_detail_String(b *testing.B) {
```

Comment (kevindiu) [reply internal/info/info_bench_test.go:51]: BTW the type name is `Detail` not `detail` 🤔
https://github.com/vdaas/vald/blob/master/internal/info/info.go#L177 

Also in unit test, the test case name is `TestStackTrace_String`.
https://github.com/vdaas/vald/blob/master/internal/info/info_test.go#L1363

Which one do you think is correct?

File: internal/info/info_bench_test.go (hunk 16 above)
Comment (kevindiu) [reply internal/info/info_bench_test.go:103]: same as below comment.

File: internal/info/info_bench_test.go (hunk 17 above)
Comment (hlts2) [reply internal/info/info_bench_test.go:51]: I think it would be nice to match it with gotests generation.

eg) if you implements the following code

```go
type detail struct{}

func (d *detail) String() {..}

type Detail struct{}

func (d *Detail) String() {}

func strings() {}

func String() {}
```

↓　the following code are generated by gotests.

```go
func Test_detail_String(t *testing.T) {
  ...
}

func TestDetail_String(t *testing.T) {
 ...
}

func Test_strings(t *testing.T) {
 ...
}

func TestString(t *testing.T) {
 ...
}
```

So, Considering the above generations, I think `BenchmarkDetail_String`  is correct in this case.
what do you think?

File: internal/info/info_bench_test.go (hunk 16 above)
Comment (hlts2) [reply internal/info/info_bench_test.go:103]: https://github.com/vdaas/vald/pull/1093/files#r603014466

File: internal/info/info_bench_test.go (hunk 17 above)
Comment (kevindiu) [reply internal/info/info_bench_test.go:51]: @vankichi what do you think?

**Review by vankichi — DISMISSED** [review]

**Review by hlts2 — DISMISSED** [review]
LGTM
